src/pieces/assets/assets_identifiers_ws.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+import pieces_os_client as pos_client
+import websocket
+from pieces.settings import Settings
+import threading
+
+class AssetsIdentifiersWS:
+    def __new__(cls,*args,**kwargs):
+        if not hasattr(cls, 'instance'):
+            cls.instance = super(AssetsIdentifiersWS, cls).__new__(cls)
+        return cls.instance
+    
+    def __init__(self, on_message_callback):
+        self.ws = None
+        self.is_connected = False
+        self.on_message_callback = on_message_callback
+
+        # Create a new event loop
+        self.loop = asyncio.new_event_loop()
+
+        # Run the event loop in a new thread
+        self.thread = threading.Thread(target=self.start_event_loop, args=(self.loop,))
+        self.thread.start()
+
+        # Open the websocket connection
+        asyncio.run_coroutine_threadsafe(self.open_websocket(), self.loop)
+
+    def start_event_loop(self, loop):
+        asyncio.set_event_loop(loop)
+        loop.run_forever()
+
+    async def open_websocket(self):
```

Comment:
since opening a websocket takes a certain amount of time, this method can technically get called many times on startup, leading to multiple websocket connections being opened. One way that could help prevent this is by making these connection management functions private to its class.

Reply:
Why the startup function will be called multiple times and even if we have a check where it make sure that it is not connected

Reply:
> Why the startup function will be called multiple times and even if we have a check where it make sure that it is not connected

We do have a check to see it's connected however this boolean will not flip until the websocket has fully connected ( which takes some time ).

Fixing this will prevent an issue from happening with this in the future.